conda_forge_tick/auto_tick.py

```diff
@@ -0,0 +1,736 @@
+
+    Parameters
+    ----------
+    feedstock_ctx: FeedstockContext
+        The node attributes
+    migrator: Migrator instance
+        The migrator to run on the feedstock
+    protocol : str, optional
+        The git protocol to use, defaults to ``ssh``
+    pull_request : bool, optional
+        If true issue pull request, defaults to true
+    rerender : bool
+        Whether to rerender
+    fork : bool
+        If true create a fork, defaults to true
+    gh : github3.GitHub instance, optional
+        Object for communicating with GitHub, if None build from $USERNAME
+        and $PASSWORD, defaults to None
+    kwargs: dict
+        The key word arguments to pass to the migrator
+
+    Returns
+    -------
+    migrate_return: MigrationUidTypedDict
+        The migration return dict used for tracking finished migrations
+    pr_json: dict
+        The PR json object for recreating the PR as needed
+
+    """
+    # get the repo
+    # TODO: stop doing this.
+    migrator.attrs = feedstock_ctx.attrs  # type: ignore
+
+    # TODO: run this in parallel
+    feedstock_dir, repo = get_repo(
+        ctx=migrator.ctx.session,
+        fctx=feedstock_ctx,
+        branch=migrator.remote_branch(feedstock_ctx),
+        feedstock=feedstock_ctx.feedstock_name,
+        protocol=protocol,
+        pull_request=pull_request,
+        fork=fork,
+    )
+
+    recipe_dir = os.path.join(feedstock_dir, "recipe")
+
+    # migrate the `meta.yaml`
+    migrate_return = migrator.migrate(recipe_dir, feedstock_ctx.attrs, **kwargs)
+    if not migrate_return:
+        logger.critical(
+            "Failed to migrate %s, %s",
+            feedstock_ctx.package_name,
+            feedstock_ctx.attrs.get("bad"),
+        )
+        eval_xonsh(f"rm -rf {feedstock_dir}")
+        return False, False
+
+    # rerender, maybe
+    diffed_files: typing.List[str] = []
+    with indir(feedstock_dir), env.swap(RAISE_SUBPROC_ERROR=False):
+        msg = migrator.commit_message(feedstock_ctx)
+        eval_xonsh("git commit -am @(msg)")
+        if rerender:
+            head_ref = eval_xonsh("git rev-parse HEAD")
+            logger.info("Rerendering the feedstock")
+            eval_xonsh("conda smithy rerender -c auto")
+            # If we tried to run the MigrationYaml and rerender did nothing (we only
+            # bumped the build number and dropped a yaml file in migrations) bail
+            # for instance platform specific migrations
+            gdiff = eval_xonsh("git diff --name-only @(head_ref)...HEAD")
+
+            diffed_files = [
+                _
+                for _ in gdiff.split()
+                if not (
+                    _.startswith("recipe")
+                    or _.startswith("migrators")
+                    or _.startswith("README")
+                )
+            ]
+
+    # TODO: Better annotation here
+    pr_json: typing.Union[MutableMapping, None, bool]
+    if isinstance(migrator, MigrationYaml) and not diffed_files:
```

Comment:
```suggestion
    if isinstance(migrator, GraphMigrator) and not diffed_files:
```

Reply:
Although I'm not super happy about this approach, maybe if you require a rerender then we check if you actually changed anything? That way we are checking an attribute rather than a class.

Reply:
Can we make a follow-on isssue for this change.  In master this applies only to the yaml based ones atm.

Reply:
#731